### PR TITLE
`configure` primes the config with data from the deployment context

### PR DESCRIPTION
**Pull request by achambers**

### CODE REVIEW

File: tests/unit/index-nodetest.js [hunk 1]
```diff
@@ -42,11 +50,41 @@ describe('the index', function() {
...
+          },
+          commandLineArgs: {
+            revision: 'some-revision'
           }
         }
       };
 
       return assert.isFulfilled(plugin.configure.call(plugin, context))
+        .then(function(result) {
+          assert.equal(result.revision, 'some-revision');
```
Comment (lukemelia) [inline tests/unit/index-nodetest.js]: It seems weird that we are asserting that the hook's promise fulfills with the config object. I didn't think that was part of the pipeline contract.

Comment (achambers) [inline tests/unit/index-nodetest.js]: It doesn't fulfil with the config object....If a hook returns an object it will be merged back in to the context....So, if the commandLineArgs contains the revision, the `configure` hook will be merging it into the context which is where the `activate` hook will reference it from.  Isn't that how we were discussing it would work?

Comment (lukemelia) [inline tests/unit/index-nodetest.js]: Ah, I understand the intent now. This is not how I envisioned this. The configure hook should be about validating config and applying defaults. In this case, the default being applied comes from the command line args.

The return values that are merged into context should be reserved for new data being produced by the pipeline.

So, I think the issue here is that we don't actually want to configure the revision key plug-in with the command line args, we want to configure the thing that will use the revision key, the redis plugin. So the configure hook of `ember-cli-deploy-redis` would apply a default revisionKey config function; something like:

``` js
function(context){
  return context.deployment.revisionKey || context.deployment.commandLineArgs["revision"];
}
```

File: index.js [hunk 2]
```diff
@@ -44,17 +44,14 @@ module.exports = {
...
-          });
+        return this._resolvePipelineData(config, context)
+          .then(validateConfig.bind(this, ui, config));
       },
 
       didBuild: function(context) {
         var deployment = context.deployment;
         var ui         = deployment.ui;
-        var config     = deployment.config[this.name] || {};
+        var config     = deployment.config[this.name] = deployment.config[this.name] || {};
```
Comment (lukemelia) [inline index.js:54]: I really hate making every plugin author do this.

File: lib/key-generators/file-hash.js [hunk 3]
```diff
@@ -33,5 +33,13 @@ module.exports = CoreObject.extend({
         fingerprint = new Fingerprint();
         return fingerprint.hashFn(contents.toString());
       })
+  },
+
+  _resolveConfigValue: function(key, config, context) {
```
Comment (lukemelia) [inline lib/key-generators/file-hash.js:38]: This should become a method on the config or context object

File: index.js (hunk 2 above)
Comment (achambers) [inline index.js:54]: Same here.....My plan is to target that sort of stuff now, once the `activate` is done.....Now that we have 3 plugins doing a similar thing I have a clearer picture about what makes sense to pull out.  So I'd like to do that in a separate PR.  Just starting looking at it now.

File: lib/key-generators/file-hash.js (hunk 3 above)
Comment (achambers) [inline lib/key-generators/file-hash.js:38]: Yep, as above....That's the plan for what I'm about starting work on @lukemelia
